### ISSUE TITLE
Non-overlapping identity check when left operand is not optional and right operand is None

### ISSUE THREAD

**Issue by sami-evangelista** [python/mypy#11888]
mypy does not complain on line 4:
```python
i: int = 1
s: str = 'hi'
i is s  # error: Non-overlapping identity check (left operand type: "int", right operand type: "str")
if i is None: # ok
   ...
```
would it be possible to raise an error in that case?

mypy version:
```
$ mypy --version
mypy 0.930
```

**Comment by chnmasta05** [comment]
I suggest adding the `topic-overlap` label.

It was decided when --strict-equality was implemented that (according to #6730) "X and None are overlapping even in strict-optional mode".

Current documentation reads:
```
* X and None are overlapping even in strict-optional mode. This is to allow
'assert x is not None' for x defined as 'x = None  # type: str' in class body
(otherwise mypy itself would have couple dozen errors because of this).
```

**Comment by gschaffner** [comment]
Is this the same as #18386?
